src/sentry/api/paginator.py

```diff
@@ -220,3 +223,95 @@ def get_result(self, limit=100, cursor=None):
             next=next_cursor,
             prev=prev_cursor,
         )
+
+
+def reverse_bisect_left(a, x, lo=0, hi=None):
+    """\
+    Similar to ``bisect.bisect_left``, but expects the data in the array ``a``
+    to be provided in descending order, rather than the ascending order assumed
+    by ``bisect_left``.
+
+    The returned index ``i`` partitions the array ``a`` into two halves so that:
+
+    - left side: ``all(val > x for val in a[lo:i])``
+    - right side: ``all(val <= x for val in a[i:hi])``
+    """
+    if lo < 0:
+        raise ValueError('lo must be non-negative')
+
+    if hi is None:
+        hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if a[mid] > x:
+            lo = mid + 1
+        else:
+            hi = mid
+
+    return lo
+
+
+class SequencePaginator(object):
+    def __init__(self, data, reverse=False):
+        self.scores, self.values = map(
+            list,
+            zip(*sorted(data, reverse=reverse)),
+        ) if data else ([], [])
+        self.reverse = reverse
+        self.search = functools.partial(
+            bisect.bisect_left if not reverse else reverse_bisect_left,
+            self.scores,
+        )
+
+    def get_result(self, limit, cursor=None):
+        if cursor is None:
+            cursor = Cursor(0, 0, False)
+
+        assert cursor.offset > -1
+
+        if cursor.value == 0:
+            position = 0 if not cursor.is_prev else len(self.scores)
+        else:
+            position = self.search(cursor.value)
+
+        position = position + cursor.offset
+
+        if not cursor.is_prev:
+            lo = max(position, 0)
+            hi = min(lo + limit, len(self.scores))
+        else:
+            # TODO: It might make sense to ensure that this hi value is at
+            # least the length of the page + 1 if we want to ensure we return a
+            # full page of results when paginating backwards while data is
+            # being mutated.
+            hi = min(position, len(self.scores))
+            lo = max(hi - limit, 0)
+
+        if self.scores:
+            prev_score = self.scores[min(lo, len(self.scores) - 1)]
+            prev_cursor = Cursor(
+                prev_score,
+                lo - self.search(prev_score, hi=lo),
+                True,
+                True if lo > 0 else False,
+            )
+
+            next_score = self.scores[min(hi, len(self.scores) - 1)]
+            next_cursor = Cursor(
+                next_score,
+                hi - self.search(next_score, hi=hi),
+                False,
+                True if hi < len(self.scores) else False,
+            )
+        else:
+            prev_cursor = Cursor(cursor.value, cursor.offset, True, False)
```

Comment:
I haven't thought through all the implications of this, and all the `hi` / `lo` math, but would it work if you always sort the scores list in ascending order (even when reverse=True), and have a bit at the end that swaps the cursors like:

```
if self.reverse:
    next_cursor, prev_cursor = prev_cursor, next_cursor
```

Reply:
I think something like this is possible, but just swapping the cursors like this:

```diff
diff --git a/src/sentry/api/paginator.py b/src/sentry/api/paginator.py
index 047e5a9..88519b8 100644
--- a/src/sentry/api/paginator.py
+++ b/src/sentry/api/paginator.py
@@ -256,13 +256,10 @@ class SequencePaginator(object):
     def __init__(self, data, reverse=False):
         self.scores, self.values = map(
             list,
-            zip(*sorted(data, reverse=reverse)),
+            zip(*sorted(data)),
         ) if data else ([], [])
         self.reverse = reverse
-        self.search = functools.partial(
-            bisect.bisect_left if not reverse else reverse_bisect_left,
-            self.scores,
-        )
+        self.search = functools.partial(bisect.bisect_left, self.scores)
 
     def get_result(self, limit, cursor=None):
         if cursor is None:
@@ -308,6 +305,10 @@ class SequencePaginator(object):
             prev_cursor = Cursor(cursor.value, cursor.offset, True, False)
             next_cursor = Cursor(cursor.value, cursor.offset, False, False)
 
+        if self.reverse:
+            prev_cursor = Cursor(next_cursor.value, next_cursor.offset, not next_cursor.is_prev, next_cursor.has_results)
+            next_cursor = Cursor(prev_cursor.value, prev_cursor.offset, not prev_cursor.is_prev, prev_cursor.has_results)
+
         return CursorResult(
             self.values[lo:hi],
             prev=prev_cursor,
```

… resulted in test failures (attached below.) I think it's probably more complex than this, but I haven't spent much time thinking about why. I'm open to coming back to it later if it's possible that it'd simplify this implementation but I'm reasonably confident this implementation works as expected and am a little hesitant to invest too much time rethinking it.

<details>

```
22:26:35 [INFO] sentry.utils.raven.SentryInternalClient: Raven is not configured (logging is disabled). Please see the documentation for more information.
============================= test session starts ==============================
platform darwin -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
rootdir: /Users/ted/workspaces/sentry/sentry, inifile: setup.cfg
plugins: xdist-1.18.2, timeout-0.5, html-1.9.0, django-2.9.1, cov-2.5.1
collected 15 items

tests/sentry/api/test_paginator.py ..FF..

=================================== FAILURES ===================================
__________ SequencePaginatorTestCase.test_descending_repeated_scores ___________
Traceback (most recent call last):
  File "/Users/ted/workspaces/sentry/sentry/tests/sentry/api/test_paginator.py", line 381, in test_descending_repeated_scores
    assert list(result) == [9, 8, 7, 6, 5]
AssertionError: assert [0, 1, 2, 3, 4] == [9, 8, 7, 6, 5]
  At index 0 diff: 0 != 9
  Use -v to get the full diff
_______________ SequencePaginatorTestCase.test_descending_simple _______________
Traceback (most recent call last):
  File "/Users/ted/workspaces/sentry/sentry/tests/sentry/api/test_paginator.py", line 335, in test_descending_simple
    assert list(result) == [9, 8, 7, 6, 5]
AssertionError: assert [0, 1, 2, 3, 4] == [9, 8, 7, 6, 5]
  At index 0 diff: 0 != 9
  Use -v to get the full diff
=============================== warnings summary ===============================
:68
  'pytest_runtest_protocol' hook uses deprecated __multicall__ argument

-- Docs: http://doc.pytest.org/en/latest/warnings.html
============================== 9 tests deselected ==============================
========= 2 failed, 4 passed, 9 deselected, 1 warnings in 0.10 seconds =========
```
</details>